### ISSUE TITLE
tf.pad() or padding="SAME" ?

### ISSUE THREAD

**Issue by zackrdavis** [distillpub/post--deconv-checkerboard#4]
You recommend `tf.pad()` before ` tf.nn.conv2d()`. I'm probably misunderstanding, but it seems that this would be handled automatically by ` tf.nn.conv2d( ... padding='SAME')`.

Here's my attempt to implement with [DCGAN-tensorflow](https://github.com/carpedm20/DCGAN-tensorflow). Many thanks if you can help, but either way, thanks for the great articles!
```
def deconv2d(input_, output_shape, k_h=5, k_w=5, d_h=2, d_w=2, stddev=0.02, name="deconv2d", with_w=False):

    with tf.variable_scope(name):
        w = tf.get_variable('w', [k_h, k_h, input_.get_shape()[-1], output_shape[-1]],
             initializer=tf.random_normal_initializer(stddev=stddev))

        resize_up = tf.image.resize_nearest_neighbor(input_, [output_shape[1], output_shape[2]])
        conv = tf.nn.conv2d(resize_up, w, strides=[1, 1, 1, 1], padding='SAME')

        biases = tf.get_variable('biases', [output_shape[-1]], initializer=tf.constant_initializer(0.0))
        conv = tf.reshape(tf.nn.bias_add(conv, biases), output_shape)
    
        if with_w:
            return conv, w, biases
        else:
            return conv
```

**Comment by colah** [comment]
Great question!

`tf.con2d(... padding='SAME')` is the same as padding with zeros using `tf.pad`. However, we'd prefer to avoid padding with zeros because it makes things unbalanced at the boundary and can cause artifacts.

I recommend using `tf.pad(..., mode='SYMMETRIC')` followed by `tf.conv2d(..., padding='VALID')` to avoid boundary effects.

**Comment by zackrdavis** [comment]
Thanks! This fixed some border artifacts that were evident even before training:

![random_unpadded_same](https://cloud.githubusercontent.com/assets/735430/21111847/6c9c41a0-c071-11e6-94bd-f23793152c16.png)

So I added:
```
padded = tf.pad(up, [[0, 0], [2, 2], [2, 2], [0, 0]], mode="SYMMETRIC")
conv = tf.nn.conv2d(padded, w, strides=[1, 1, 1, 1], padding='VALID')
```

and got:
![random_padded_symmetric](https://cloud.githubusercontent.com/assets/735430/21111842/6ac3ed10-c071-11e6-93da-ad853e279f6e.png)

**Comment by colah** [comment]
Awesome! Good luck, and let me know if you run into any problems!
